### PR TITLE
ADDED DRL ALGORITHN HANDLER.

### PR DESCRIPTION
**Pull request by GilLifshits**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -32,11 +32,8 @@
         ONLY_INFERENCE=False,
         EPOCHS=50,
         ROLE='Car1',
-        EXPLORATION_EXPLOTATION_THRESHOLD=2500,
-        TIME_BETWEEN_STEPS=0.05,
-        LEARNING_RATE=0.0000001,
-        N_STEPS=200,
-        BATCH_SIZE=200,
+        EXPLORATION_EXPLOTATION_THRESHOLD=50,
+        MODEL_TYPE='A2C',
```
Comment (GilLifshits) [inline main.py]: Make 'A2C' as param in config as ENUM, and use it.

File: main.py [hunk 2]
```diff
@@ -59,16 +56,10 @@
...
-    #     TIME_BETWEEN_STEPS=0.05,
-    #     LEARNING_RATE=1,
-    #     N_STEPS=160,
-    #     BATCH_SIZE=160,
-    #     FIXED_THROTTLE=0.75,
-    #     THROTTLE_FAST=1,
-    #     THROTTLE_SLOW=0.5,
-    #     LOAD_WEIGHT_DIRECTORY= get_latest_model("experiments")
+    #     LOAD_WEIGHT_DIRECTORY= get_latest_model("experiments"),
+    #     MODEL_TYPE='DQN'
```
Comment (GilLifshits) [inline main.py]: Same as comment above

File: main.py [hunk 3]
```diff
@@ -59,16 +56,10 @@
...
     #     EXPLORATION_EXPLOTATION_THRESHOLD=0,
-    #     TIME_BETWEEN_STEPS=0.05,
-    #     LEARNING_RATE=1,
-    #     N_STEPS=160,
-    #     BATCH_SIZE=160,
-    #     FIXED_THROTTLE=0.75,
-    #     THROTTLE_FAST=1,
-    #     THROTTLE_SLOW=0.5,
-    #     LOAD_WEIGHT_DIRECTORY= get_latest_model("experiments")
+    #     LOAD_WEIGHT_DIRECTORY= get_latest_model("experiments"),
```
Comment (GilLifshits) [inline main.py]: Change get_latest_model to a function that specifies the model files you want to load, and not the latest.

File: main.py [hunk 4]
```diff
@@ -59,16 +56,10 @@
     #     EXPERIMENT_ID='Experiment2_INFERENCE_ONLY',
     #     ONLY_INFERENCE=True,
     #     EPOCHS=5,
-    #     ROLE='Both',
+    #     ROLE='Car1',
```
Comment (GilLifshits) [inline main.py]: Define role in Config and use it here

File: utils/drl_algorithm_selector.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from stable_baselines3 import PPO, DQN, A2C
+
+class ModelSelector:
+    def __init__(self, model_type, env, config):
+        self.model_type = model_type
+        self.env = env
+        self.config = config
+        self.model = self.init_model()
+
+    def init_model(self):
```
Comment (KathyAgafonov) [inline utils/drl_algorithm_selector.py:10]: Consider refactoring this method to be more modular by using helper methods or a dictionary to map `model_type` to configurations and constructors.

File: utils/drl_algorithm_selector.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+from stable_baselines3 import PPO, DQN, A2C
+
+class ModelSelector:
+    def __init__(self, model_type, env, config):
+        self.model_type = model_type
```
Comment (KathyAgafonov) [inline utils/drl_algorithm_selector.py]: Consider refactoring this method to be more modular by using helper methods or a dictionary to map 'model_type' to configurations and constructors.

File: utils/experiment.py [hunk 7]
```diff
@@ -21,7 +21,44 @@ class Experiment:
...
+    ENT_COEF: float = 0.0
+    VF_COEF: float = 0.5
+    MAX_GRAD_NORM: float = 0.5
+    USE_SDE: bool = False
+    SDE_SAMPLE_FREQ: int = -1
+
+    # Model specific configurations (e.g., custom configs per model)
+    MODEL_CONFIGS: Dict[str, Dict] = field(default_factory=dict)
+
+    def __post_init__(self):
```
Comment (KathyAgafonov) [inline utils/experiment.py]: There are two 'post_init' methods.

File: utils/experiment.py [hunk 8]
```diff
@@ -21,7 +21,44 @@ class Experiment:
...
+        # Default configurations per model type
+        self.MODEL_CONFIGS = {
```
Comment (KathyAgafonov) [inline utils/experiment.py]: Consider separating the default 'MODEL_CONFIGS' into a external config file.

File: utils/training_loop.py [hunk 9]
```diff
@@ -13,42 +13,47 @@ def training_loop(experiment, env, agent, model):
...
+        all_rewards, all_actions = [], []
+
+        for episode in range(experiment.EPOCHS):
+            print(f"@ Episode {episode + 1} @")
+            current_state = env.reset()
+            done = False
+            episode_sum_of_rewards, steps_counter = 0, 0
+            episode_counter += 1
+            resume_experiment_simulation(env)
+            actions_per_episode = []
```
Comment (KathyAgafonov) [inline utils/training_loop.py:27]: You define 'actions_per_episode' but never append actions to it.

File: utils/training_loop.py [hunk 10]
```diff
@@ -13,42 +13,47 @@ def training_loop(experiment, env, agent, model):
         print('Only Inference')
```
Comment (KathyAgafonov) [inline utils/training_loop.py:13]: Add documentation explaining the function 'training_loop'.

File: utils/drl_algorithm_selector.py (hunk 6 above)
Comment (GilLifshits) [reply utils/drl_algorithm_selector.py]: For example:
class ModelType(Enum):
    PPO = 'PPO'
    DQN = 'DQN'
    A2C = 'A2C'
    
   Make sure it is defined outside this scope, and change main file to define model type from this enum as well

File: utils/training_loop.py [hunk 11]
```diff
@@ -13,42 +13,47 @@ def training_loop(experiment, env, agent, model):
...
         model = PPO.load(experiment.LOAD_WEIGHT_DIRECTORY)
```
Comment (GilLifshits) [inline utils/training_loop.py]: The inference needs to be according to the model that we trained on (PPO/A2C/DQN)

File: utils/training_loop.py [hunk 12]
```diff
@@ -59,14 +64,17 @@ def plot_results(experiment, all_rewards, all_actions):
...
     airsim_manager = AirsimManager(experiment)
     env = DummyVecEnv([lambda: Agent(experiment, airsim_manager)])
     agent = Agent(experiment, airsim_manager)
-    model = PPO(policy=PPO_MLP_Policy, env=env, verbose=1,
-                learning_rate=experiment.LEARNING_RATE,
-                n_steps=experiment.N_STEPS,
-                batch_size=experiment.BATCH_SIZE)
+    # Initialize model selector with the desired type and configuration
+    model_selector = ModelSelector(experiment.MODEL_TYPE, env, config={
+        'learning_rate': experiment.LEARNING_RATE,
```
Comment (GilLifshits) [inline utils/training_loop.py]: This is not supposed to be defined here

File: utils/training_loop.py [hunk 13]
```diff
@@ -59,14 +64,17 @@ def plot_results(experiment, all_rewards, all_actions):
...
+        'n_steps': experiment.N_STEPS,
+        'batch_size': experiment.BATCH_SIZE
+    })
+    model = model_selector.get_model()
+    logger = configure(experiment.EXPERIMENT_PATH, ["stdout", "csv", "tensorboard"])
     model.set_logger(logger)
 
     model, collision_counter, all_rewards, all_actions = training_loop(experiment=experiment, env=env, agent=agent,
```
Comment (GilLifshits) [inline utils/training_loop.py]: In the line of 
def resume_experiment_simulation(env):
Complete the TODO (why not using resume_simulation from airsim manager?)

**Comment by GilLifshits** [comment]
I found another bug, that when defining 2 experiment in main:
experiments = [experiment1, experiment2]
    for experiment in experiments:
    
experiment2 starts in init values that are incorrect (right beside each other), casing the experiment to fail.

File: main.py (hunk 3 above)
Comment (arielharush96) [reply main.py]: there is another function for that : get_model_from_specific_directory. i just prefer to use the function in this code.

File: main.py (hunk 4 above)
Comment (arielharush96) [reply main.py]: already in experiment :     # Model and Training Configuration
    ROLE: str = ""  # Which car is using the DRL model. Car1, Car2,Both

File: utils/experiment.py (hunk 7 above)
Comment (arielharush96) [reply utils/experiment.py]: @GilLifshits method

File: utils/training_loop.py (hunk 9 above)
Comment (arielharush96) [reply utils/training_loop.py:27]: it was for debugging, deleted

File: main.py (hunk 1 above)
Comment (ariel-harush-sndk) [reply main.py]: Done

File: main.py (hunk 2 above)
Comment (ariel-harush-sndk) [reply main.py]: Done

File: main.py (hunk 4 above)
Comment (arielharush96) [reply main.py]: Done

File: utils/drl_algorithm_selector.py (hunk 6 above)
Comment (arielharush96) [reply utils/drl_algorithm_selector.py]: Done

File: utils/experiment.py (hunk 8 above)
Comment (arielharush96) [reply utils/experiment.py]: Done

File: utils/training_loop.py (hunk 10 above)
Comment (arielharush96) [reply utils/training_loop.py:13]: Done

File: utils/training_loop.py (hunk 11 above)
Comment (arielharush96) [reply utils/training_loop.py]: Done

File: utils/training_loop.py (hunk 12 above)
Comment (arielharush96) [reply utils/training_loop.py]: Done

File: utils/training_loop.py (hunk 13 above)
Comment (arielharush96) [reply utils/training_loop.py]: Done

**Pull request merged**